gooddata-sdk/gooddata_sdk/catalog/data_source/declarative_model/data_source.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import annotations
+
+from typing import Any, Optional
+
+from gooddata_metadata_client.model.declarative_data_source import DeclarativeDataSource
+from gooddata_metadata_client.model.declarative_data_sources import DeclarativeDataSources
+from gooddata_sdk.catalog.data_source.declarative_model.physical_model.pdm import CatalogDeclarativePdm
+from gooddata_sdk.catalog.entity import CatalogTypeEntity
+
+
+class CatalogDeclarativeDataSources:
+    def __init__(self, data_sources: list[CatalogDeclarativeDataSource]):
+        self.data_sources = data_sources
+
+    @classmethod
+    def from_api(cls, entity: dict[str, Any]) -> CatalogDeclarativeDataSources:
+        data_sources = [CatalogDeclarativeDataSource.from_api(v) for v in entity["data_sources"]]
+        return cls(data_sources)
+
+    def to_api(self) -> DeclarativeDataSources:
+        return DeclarativeDataSources(data_sources=[v.to_api() for v in self.data_sources])
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, CatalogDeclarativeDataSources):
+            return False
+        return self.data_sources == other.data_sources
+
+
+class CatalogDeclarativeDataSource(CatalogTypeEntity):
+    def __init__(
+        self,
+        id: str,
+        type: str,
+        name: str,
+        url: str,
+        enable_caching: bool,
+        pdm: CatalogDeclarativePdm,
+        schema: Optional[str],
+        cache_path: Optional[list[str]] = None,
+    ):
+        super(CatalogDeclarativeDataSource, self).__init__(id, type)
+        self.schema = schema
+        self.enable_caching = enable_caching
+        self.cache_path = cache_path
+        self.name = name
+        self.url = url
+        self.pdm = pdm
+
+    @classmethod
+    def from_api(cls, entity: dict[str, Any]) -> CatalogDeclarativeDataSource:
+        return cls(
+            id=entity["id"],
+            name=entity["name"],
+            url=entity["url"],
+            type=entity["type"],
+            enable_caching=entity["enable_caching"],
+            pdm=CatalogDeclarativePdm.from_api(entity),
+            schema=entity.get("schema"),
+            cache_path=entity.get("cache_path"),
+        )
+
+    def to_api(
+        self,
+        user: Optional[str] = None,
+        password: Optional[str] = None,
+        token: Optional[str] = None,
+    ) -> DeclarativeDataSource:
+        kwargs: dict[str, Any] = {}
+        if self.schema:
+            kwargs["schema"] = self.schema
+        if self.cache_path:
+            kwargs["cache_path"] = self.cache_path
+        if user:
+            kwargs["user"] = user
+        if password:
+            kwargs["password"] = password
+        if token:
+            kwargs["token"] = token
+        return DeclarativeDataSource(
+            id=self.id,
+            name=self.name,
+            url=self.url,
+            type=self.type,
+            enable_caching=self.enable_caching,
+            pdm=self.pdm.to_api(),
```

Comment:
pdm should be part of kwargs I think...

Reply:
It is not optional, so I think it should not. Why you think it should?

Reply:
ok, I realized that it is optional

Reply:
The chain data_source -> pdm -> tables refactored, so data_source.pdm.pdm.tables access pattern is not needed.
Optionality fixed.